compiler/dyno/lib/uast/chpl-syntax-printer.cpp

```diff
@@ -0,0 +1,1065 @@
+      case Variable::Kind::CONST_REF: return "const ref";
+      case Variable::Kind::INDEX: assert(false);
+    }
+    assert(false);
+    return "";
+  }
+
+  static const char* kindToString(VisibilityClause::LimitationKind kind) {
+    switch (kind) {
+      case VisibilityClause::LimitationKind::NONE: assert(false);
+      case VisibilityClause::LimitationKind::ONLY: return "only";
+      case VisibilityClause::LimitationKind::EXCEPT: return "except";
+      case VisibilityClause::LimitationKind::BRACES: assert(false);
+      default: return "";
+    }
+  }
+
+
+  /** visit each elt of begin..end, outputting `separator` between each.
+   * `surroundBegin` and `surroundEnd` are output before and after respectively
+   * if not null */
+  template<typename It>
+  void interpose(std::ostream& os, It begin, It end, const char* separator, const char* surroundBegin=nullptr, const char* surroundEnd=nullptr) {
+    bool first = true;
+    if (surroundBegin) os << surroundBegin;
+    for (auto it = begin; it != end; it++) {
+      if (!first) os << separator;
+      printAst(os,*it);
+      first = false;
+    }
+    if (surroundEnd) os << surroundEnd;
+  }
+
+  template<typename T>
+  void interpose(std::ostream& os, T xs, const char* separator, const char* surroundBegin=nullptr, const char* surroundEnd=nullptr) {
+    interpose(os, xs.begin(), xs.end(), separator, surroundBegin, surroundEnd);
+  }
+
+
+  /*
+   * helper for printing descendants of simpleBlockLike to handle when to print
+   * the optional opening keyword and if braces should follow that keyword
+   */
+  template<typename It>
+  void printBlockWithStyle(std::ostream& os, BlockStyle style,  It begin, It end, const char* implicitOpeningKeyword=nullptr) {
+    if (implicitOpeningKeyword && (style == BlockStyle::IMPLICIT || style == uast::BlockStyle::UNNECESSARY_KEYWORD_AND_BLOCK)) {
+      os << implicitOpeningKeyword;
+    }
+    if (style != BlockStyle::IMPLICIT) {
+      interpose(os, begin, end, "\n", "{","}");
+    } else {
+      interpose(os, begin, end, "\n");
+    }
+  }
+
+
+  template<typename T>
+  void printBlockWithStyle(std::ostream& os, BlockStyle style,  T xs, const char* implicitOpeningKeyword=nullptr) {
+    printBlockWithStyle(os, style, xs.begin(), xs.end(), implicitOpeningKeyword);
+  }
+
+  template<typename T>
+  void visitLiteral(std::ostream& ss_, const T* node) {
+    ss_ << node->text().c_str();
+  }
+
+  // TODO: Nesting
+  // TODO: Semicolons
+  // TODO: Newlines
+
+
+  struct ChplSyntaxVisitor {
+    std::ostringstream ss_;
+    ChplSyntaxVisitor(std::ostringstream ss_ = std::ostringstream()) {}
+
+    void printLinkages(const Decl* node) {
+      if (node->linkage() != Decl::Linkage::DEFAULT_LINKAGE) {
+        ss_ << kindToString(node->linkage()) << " ";
+        if (node->linkageName() != nullptr) {
+          printAst(ss_, node->linkageName());
+          ss_ << " ";
+        }
+      }
+    }
+
+    void visit(const uast::ASTNode* node) {
```

Comment:
I am thinking we should have this be an `assert(false && "Unhandled AST node")`.